server/src/main/java/org/elasticsearch/action/fieldcaps/FieldCapabilities.java

```diff
@@ -107,22 +117,47 @@ public void writeTo(StreamOutput out) throws IOException {
     @Override
     public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startObject();
-        builder.field("type", type);
-        builder.field("searchable", isSearchable);
-        builder.field("aggregatable", isAggregatable);
+        builder.field(TYPE_FIELD.getPreferredName(), type);
+        builder.field(SEARCHABLE_FIELD.getPreferredName(), isSearchable);
+        builder.field(AGGREGATABLE_FIELD.getPreferredName(), isAggregatable);
         if (indices != null) {
-            builder.field("indices", indices);
+            builder.field(INDICES_FIELD.getPreferredName(), indices);
         }
         if (nonSearchableIndices != null) {
-            builder.field("non_searchable_indices", nonSearchableIndices);
+            builder.field(NON_SEARCHABLE_INDICES_FIELD.getPreferredName(), nonSearchableIndices);
         }
         if (nonAggregatableIndices != null) {
-            builder.field("non_aggregatable_indices", nonAggregatableIndices);
+            builder.field(NON_AGGREGATABLE_INDICES_FIELD.getPreferredName(), nonAggregatableIndices);
         }
         builder.endObject();
         return builder;
     }
 
+    public static FieldCapabilities fromXContent(String name, XContentParser parser) throws IOException {
+        return PARSER.parse(parser, name);
+    }
+
+    @SuppressWarnings("unchecked")
+    private static ConstructingObjectParser<FieldCapabilities, String> PARSER = new ConstructingObjectParser<>(
+        "field_capabilities",
+        true,
+        (a, name) -> new FieldCapabilities(name,
+            (String) a[0],
+            (boolean) a[1],
+            (boolean) a[2],
+            a[3] != null ? ((List<String>) a[3]).toArray(new String[0]) : null,
+            a[4] != null ? ((List<String>) a[4]).toArray(new String[0]) : null,
+            a[5] != null ? ((List<String>) a[5]).toArray(new String[0]) : null));
+
+    static {
+        PARSER.declareString(ConstructingObjectParser.constructorArg(), TYPE_FIELD);
+        PARSER.declareBoolean(ConstructingObjectParser.constructorArg(), SEARCHABLE_FIELD);
+        PARSER.declareBoolean(ConstructingObjectParser.constructorArg(), AGGREGATABLE_FIELD);
+        PARSER.declareStringArray(ConstructingObjectParser.optionalConstructorArg(), INDICES_FIELD);
+        PARSER.declareStringArray(ConstructingObjectParser.optionalConstructorArg(), NON_SEARCHABLE_INDICES_FIELD);
+        PARSER.declareStringArray(ConstructingObjectParser.optionalConstructorArg(), NON_AGGREGATABLE_INDICES_FIELD);
+    }
```

Comment:
maybe we could move `FieldCapabilitiesTests` to extend `AbstractSerializingTestCase` given that we can now parse it back.